### PR TITLE
Adding LSP Features: Highlight unresolved references, add a basic onHover functionality + Misc fixes

### PR DESCRIPTION
**Pull request by OceanOak**
No changelog

 This PR:
 - Gets the LSP woking again
 - Reimplements `dangerLanguageServerProtocolReadNextMessage` in Darklang
 - Better Semantic Tokenization 
 - Adds a very basic `onHover` functionality
 - `onSave` resend or clear diagnostics
 -  Implements sending diagnostics for unresolved references
 - Updates WT2PT to return unresolved names alongside PT

### CODE REVIEW

File: packages/darklang/languageTools/lsp-server/handleIncomingMessage.dark [hunk 1]
```diff
@@ -41,6 +41,42 @@ module Darklang =
...
+      //   LanguageServerProtocol.Diagnostic.Diagnostic
+      //     { range = convertParserRangeToLspRange range
+      //       severity = Stdlib.Option.Option.Some severity
+      //       code = Stdlib.Option.Option.None
+      //       codeDescription = Stdlib.Option.Option.None
+      //       source = Stdlib.Option.Option.None
+      //       message = message
+      //       tags = Stdlib.Option.Option.None
+      //       relatedInformation = Stdlib.Option.Option.None
+      //       data = Stdlib.Option.Option.None }
```
Comment (StachuDotNet) [inline packages/darklang/languageTools/lsp-server/handleIncomingMessage.dark:77]: I'd say it's useful. as follow-ups:
- move severity to first param
- add specific fns like `createErrorDiagnostic` that just apply the severity (you might need to make the new fns temporarily take in all 3 params, but once interpreter work is merged we can tidy)

File: packages/darklang/languageTools/lsp-server/docSync.dark [hunk 2]
```diff
@@ -31,31 +31,46 @@ module Darklang =
...
+          let sourceFile =
+            match parsed with
+            | Ok parsedFile ->
+              match parsedFile with
+              | SourceFile sf -> sf
+              | _ -> log "Invalid source file"
+            | Error err -> log err
+
+          let (_ptCliScriptModule, unresolvedNames) =
+            Parser.CliScript.parse
```
Comment (OceanOak) [inline packages/darklang/languageTools/lsp-server/docSync.dark:52]: Is it ok to reuse `CliScript.parse` here?

File: packages/darklang/languageTools/writtenTypesToProgramTypes.dark [hunk 3]
```diff
@@ -1416,5 +1416,7 @@ module Darklang =
         : (ProgramTypes.SourceFile.SourceFile *
```
Comment (OceanOak) [inline packages/darklang/languageTools/writtenTypesToProgramTypes.dark:1416]: why do we have `parsedFileAsSourceFile` as a separate function, and not just as a part of the `sourceFile` function?

File: packages/darklang/languageTools/writtenTypesToProgramTypes.dark [hunk 4]
```diff
@@ -1416,5 +1416,7 @@ module Darklang =
...
           List<Parser.Range * RuntimeErrors.NameResolution.Error>)
         =
+        let pm = LanguageTools.PackageManager.pm ()
```
Comment (OceanOak) [inline packages/darklang/languageTools/writtenTypesToProgramTypes.dark:1419]: Should pm be sent as a parameter?

File: packages/darklang/languageTools/parser/cliScript.dark [hunk 5]
```diff
@@ -146,10 +158,12 @@ module Darklang =
           (pm: ProgramTypes.PackageManager.PackageManager)
           (filename: String)
```
Comment (OceanOak) [inline packages/darklang/languageTools/parser/cliScript.dark:159]: Why do we have both `scriptName` and `fileName`? is file name the full path?

File: packages/darklang/languageTools/lsp-server/docSync.dark (hunk 2 above)
Comment (StachuDotNet) [reply packages/darklang/languageTools/lsp-server/docSync.dark:52]: Yes - I think that's probably ideal, for now - just focusing on .dark "scripts." Once the LS has a virtual document system set up, where we can be "editing the contents of X package" or something else, we can adjust here accordingly.

File: packages/darklang/languageTools/writtenTypesToProgramTypes.dark (hunk 4 above)
Comment (StachuDotNet) [reply packages/darklang/languageTools/writtenTypesToProgramTypes.dark:1419]: Eventually, yes. I suspect we had just punted that, as it'll require us to trickle the `pm` object in a bunch of places.

File: packages/darklang/languageTools/writtenTypesToProgramTypes.dark (hunk 3 above)
Comment (StachuDotNet) [reply packages/darklang/languageTools/writtenTypesToProgramTypes.dark:1416]: not sure :) can probably tidy.

File: packages/darklang/languageTools/parser/cliScript.dark (hunk 5 above)
Comment (StachuDotNet) [reply packages/darklang/languageTools/parser/cliScript.dark:159]: Hm, I'm not sure. It doesn't seem like `fileName` is even used?

Comment (OceanOak) [reply packages/darklang/languageTools/parser/cliScript.dark:159]: yes, it isn't used

File: packages/darklang/languageServerProtocol/language/onHover.dark
Comment (StachuDotNet) [inline packages/darklang/languageServerProtocol/language/onHover.dark:1]: Please check out the common.dark in the directory above.
All of the LSP types should already be _somewhere_ in the `languageServerProtocol` directory -- some still in commented typescript, some 'ported' to Dark.
When adding any new types, please make sure to remove the commented typescript source, so we don't end up adding it again later.
Also, when adding those new types, check to see where it's used to see if it should stay in common.dark -- `MarkupKind` for example is used in Completions, OnHover, and SignatureHelp handshakes, so it should probably 'stay' in that file.

File: packages/darklang/languageTools/lsp-server/docSync.dark [hunk 6]
```diff
@@ -31,17 +31,46 @@ module Darklang =
                     ) }
             )
 
+        let parseAndreportDiagnostics
```
Comment (StachuDotNet) [inline packages/darklang/languageTools/lsp-server/docSync.dark:34]: picky -- `r` -> `R`

File: packages/darklang/languageTools/parser/cliScript.dark (hunk 5 above)
Comment (StachuDotNet) [reply packages/darklang/languageTools/parser/cliScript.dark:159]: probably cool to remove it, then, unless I'm missing something

File: packages/darklang/languageTools/writtenTypes.dark [hunk 7]
```diff
@@ -351,11 +351,7 @@ module Darklang =
           fields: List<Expr> *
           /// between the typeName and the caseName
           symbolDot: Range
-        | EPipeFnCall of
-          Range *
-          fnName: Name *
-          typeArgs: List<TypeReference.TypeReference> *
-          args: List<Expr>
+        | EPipeFnCall of Range * fnName: QualifiedFnIdentifier * args: List<Expr>
```
Comment (StachuDotNet) [inline packages/darklang/languageTools/writtenTypes.dark:354]: :heart:

**Review by StachuDotNet — CHANGES_REQUESTED** [review]
This is looking good. A few tiny things to fix, but otherwise I think this is OK to merge.

I'll be honest in that I haven't read WT2PT changes in full detail -- maybe once my PR is merged, though, I should sit down and do a full pass-through of the file to provide feedback+ideas. I think it's generally fine, but a round of nitpicking and chatting might be useful :)

**Pull request merged**